### PR TITLE
[TT-12840, DX-1620] Added explanation of bug fix missed in 5.4 RN

### PR DESCRIPTION
**Pull request by andyo-tyk**
DX-1620

---

### Preview Link
https://deploy-preview-5260--tyk-docs.netlify.app/docs/nightly/getting-started/key-concepts/rate-limiting/#combining-multiple-policies-configuring-rate-limits
https://deploy-preview-5260--tyk-docs.netlify.app/docs/nightly/product-stack/tyk-gateway/release-notes/version-5.4/

<br>

### Description
We missed a bug fix from the 5.4 release notes, as it was fixed as part of another change. Adding this to the release notes and also an explanation of the functionality to the main docs.

___

### **PR Type**
Documentation, Enhancement

___

### **Description**
- Added documentation explaining how Tyk combines multiple policies with rate limits, highlighting the corrected behavior in version 5.4.0.
- Updated the release notes to include details of a bug fix in the calculation of effective rate limits when multiple policies are applied to a key.
- Provided a detailed explanation of the previous incorrect behavior and the new corrected logic for rate limit calculations.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>rate-limiting.md</strong><dd><code>Document combining multiple rate limit policies and bug fix</code></dd></summary>
<hr>

tyk-docs/content/getting-started/key-concepts/rate-limiting.md

<li>Added explanation for combining multiple policies with rate limits.<br> <li> Clarified the corrected behavior for rate limit calculation.<br> <li> Included a note about the bug fix in rate limit calculation.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk-docs/pull/5260/files#diff-a0f383800801ae765ec875c3c98610f657751ec0779b9e23970357b74fb58e9a">+14/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>version-5.4.md</strong><dd><code>Update release notes with rate limit calculation fix</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tyk-docs/content/product-stack/tyk-gateway/release-notes/version-5.4.md

<li>Updated release notes to include a bug fix for rate limit calculation.<br> <li> Explained the impact of the fix on existing policies.<br> <li> Added a detailed description of the previous and corrected behavior.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk-docs/pull/5260/files#diff-3cbbef4987e7209220db6a08dd8059cab4c9df294a36572ccb1ad1b858593e2a">+16/-1</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: tyk-docs/content/product-stack/tyk-gateway/release-notes/version-5.4.md [hunk 1]
```diff
@@ -297,6 +297,21 @@ Fixed an issue where the Gateway did not respect API domain settings when there
 Addressed a problem with nested field mapping in UDG for GraphQL (GQL) operations. Previously, querying a single nested field caused an error, while including another *normal* field from the same level allowed the query to succeed. This issue has been fixed to ensure consistent behavior regardless of the query composition.
 </details>
 </li>
+<li>
+<details>
+<summary>Fixed an error in the calculation of effective rate limit from multiple policies</summary>
+
+Fixed a long-standing bug in the algorithm used to determine the effective rate limit when multiple policies are applied to a key. If more than one policy is applied to a key then Tyk will apply the highest request rate permitted by any of the policies that defines a rate limit.
+
+Rate limits in Tyk are defined using two elements: `rate`, which is the number of requests and `per`, which is the period over which those requests can be sent. So, if `rate` is 90 and `per` is 30 seconds for a key, Tyk will permit a maximum of 100 requests to be made using the key in a 30 second period, giving an effective maximum of 180 requests per minute (or 3 rps).
```
Comment (dcs3spp) [inline tyk-docs/content/product-stack/tyk-gateway/release-notes/version-5.4.md]: Should the maximum requests be 90?

File: tyk-docs/content/getting-started/key-concepts/rate-limiting.md [hunk 2]
```diff
@@ -70,6 +70,20 @@ When multiple rate limits are configured, they are assessed in this order (if ap
 2. Key-level global rate limit
 3. Key-level per-API rate limit
 
+### Combining multiple policies configuring rate limits
+
+If more than one policy defining a rate limit is applied to a key then Tyk will apply the highest request rate permitted by any of the policies that defines a rate limit.
+
+If `rate` and `per` are configured in multiple policies applied to the same key then the Gateway will determine the effective rate limit configured for each policy and apply the highest to the key.
+
+Given, policy A with `rate` set to 90 and `per` set to 30 seconds (3rps) and policy B with `rate` set to 100 and `per` set to 10 seconds (10rps). If both are applied to a key, Tyk will take the rate limit from policy B as it results in a higher effective request rate (10rps).
```
Comment (dcs3spp) [inline tyk-docs/content/getting-started/key-concepts/rate-limiting.md:79]: Optional: Could present policy A and policy B in a table

**Review by dcs3spp — CHANGES_REQUESTED** [review]
PR LGTM, not sure if there was a typo in example given in change log, should 100 requests be updated to 90 requests in 2nd paragraph of the fixed change-log item?

File: tyk-docs/content/product-stack/tyk-gateway/release-notes/version-5.4.md (hunk 1 above)
Comment (andyo-tyk) [reply tyk-docs/content/product-stack/tyk-gateway/release-notes/version-5.4.md]: Yes, good spot - I was already thinking about my example of the buggy behaviour 🤣

Comment (andyo-tyk) [reply tyk-docs/content/product-stack/tyk-gateway/release-notes/version-5.4.md]:
```suggestion
Rate limits in Tyk are defined using two elements: `rate`, which is the number of requests and `per`, which is the period over which those requests can be sent. So, if `rate` is 90 and `per` is 30 seconds for a key, Tyk will permit a maximum of 90 requests to be made using the key in a 30 second period, giving an effective maximum of 180 requests per minute (or 3 rps).
```

**Review by dcs3spp — APPROVED** [review]

File: tyk-docs/content/getting-started/key-concepts/rate-limiting.md (hunk 2 above)
Comment (andyo-tyk) [reply tyk-docs/content/getting-started/key-concepts/rate-limiting.md:79]: I considered this, but with only two lines of data, wonder whether it's overkill? Maybe something to implement in the future if there's a visual design/layout change?

**Comment by dcs3spp** [comment]
PR LGTM, @lghiur @OldStubbsy would you be able to peer review and if ok approve and then I will merge and release

File: tyk-docs/content/product-stack/tyk-gateway/release-notes/version-5.4.md [hunk 3]
```diff
@@ -297,6 +297,21 @@ Fixed an issue where the Gateway did not respect API domain settings when there
...
+
+Previously, Tyk would take the highest `rate` and the highest `per` from the policies applied to a key when determining the effective rate limit. So, if policy A had `rate` set to 90 and `per` set to 30 seconds (3rps) while policy B had `rate` set to 100 and `per` set to 10 seconds (10rps) and both were applied to a key, the rate limit configured in the key would be: `rate = 100` and `per = 30` giving a rate of 3.33rps.
+
+With the fix applied in Tyk 5.4.0, the Gateway will now apply the highest effective rate to the key - so in this example, the key would take the rate limit from policy B: `rate = 100` and `per = 10` (10rps).
+
+Note that this corrected logic will be applied when creating new keys from Tyk 5.4.0 onwards, so if you have existing policies with rate limits defined and are applying these to keys, there may be a change in the effective rate limit configured in the key.
```
Comment (andyo-tyk) [inline tyk-docs/content/product-stack/tyk-gateway/release-notes/version-5.4.md]:
```suggestion
Note that this corrected logic is applied when access keys are presented in API requests. If you are applying multiple policies to keys, there may be a change in the effective rate limit when using Tyk 5.4.0 compared with pre-5.4.0 versions.
```

**Review by OldStubbsy — APPROVED** [review]
Looks good to me. Thanks!

**Review by lghiur — APPROVED** [review]

**Comment by dcs3spp** [comment]
Thanks @lghiur @OldStubbsy for peer approval, I'll publish PR for @andyo-tyk now

**Pull request merged**